### PR TITLE
Add updating of soil moisture and soil temperature for Noah-MP to global_cycle

### LINKED ISSUE #832: Update global_cycle to handle Noah-MP soil moisture and soil temp. increments in land_increments.f90
We are targeting introduction of a soil moisture and soil temperature analysis for GFSv17. The increments will be created by the GSI (from the assimilation of screen-level and other atmos obs), and need to then be added to the model restarts using global_cycle. 

At the moment, global_cycle can add soil temperature increments to the Noah model. So we need to: 
-add in soil moisture increments 
-add the model-dependent code for Noah_MP for soil temperature increments
(when the soil temperature changes around tfreeze, we need to adjust the soil moisture frozen fraction). 

After discussion with @barlage , we decided to do this in two steps: 
1. Add in the soil moisture increments, and Noah-MP updates, but limit updates (for both soil moisture and soil temperature) to only apply to non-frozen soils, thus avoiding the need to insert model physics to adjust frozen soil moisture content following temperature changes. 

2. Mike's team works on the most appropriate way to deal with frozen conditions, so that we can later also apply soil temperature increments when below freezing (frozen soil moisture has little interaction with the atmosphere, so we'll likely keep the non-frozen QC for soil moisture.

### PR DESCRIPTION
**Pull request by ClaraDraper-NOAA**
## DESCRIPTION OF CHANGES: 

Main changes: 
- Added option to update Noah-MP restarts with increments from the gsi to global_cycle/land_increments 
- Added option to update soil moisture for Noah-MP only (previously was soil temperature only) 
- Removed LSM option from namelist, since was redundant with IS_NOAHMP 
- Replaced the C768.landincsoil test (STC update to Noah restarts) with the C192.landincsoilnoahmp test (SLC, SMC, and STC updates to Noah-MP restarts) 

Inputs for new test are on hera:
/scratch2/BMC/gsienkf/Clara.Draper/global_cycle_newtest/input_data_noahmp/ 

.../input_data/sfcincr_gsi can be deleted (everything else in that directory is still needed) 

Example output for new test on hera: 
/scratch2/BMC/gsienkf/Clara.Draper/global_cycle_newtest/baseline_data/c192.lndincsoilnoahmp

## TESTS CONDUCTED: 
All tests used abe5c94.

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera and WCOSS2).
- [x] Compile branch on Hera using GNU.
- [x] Compile branch in 'Debug' mode on WCOSS2.
- [x] Run unit tests locally on any Tier 1 machine. (Done on Hera and WCOSS2. All passed.)
- [x] Run relevant consistency tests locally on all Tier 1 machine.

Also created the Doxygen on Hera. No warnings were noted. 

## DEPENDENCIES:
None 

## DOCUMENTATION:
None. 

## ISSUE: 
Fixes #832

## CONTRIBUTORS: 
None.

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
@ClaraDraper-NOAA I will look at this today. I can help with testing on machines you can't access.

**Comment by GeorgeGayno-NOAA** [comment]
@WalterKolczynski-NOAA - This PR changes the global_cycle program's NAMCYC namelist. The LSM variable is being removed from that namelist.  Does this change affect the global workflow?

**Comment by WalterKolczynski-NOAA** [comment]
@GeorgeGayno-NOAA I don't think so. We're still using the `global_cycle.sh` from ufs-utils which already seems to not include the setting in its namelist.

File: reg_tests/global_cycle/C192.lndincsoilnoahmp.sh
Comment (GeorgeGayno-NOAA) [inline reg_tests/global_cycle/C192.lndincsoilnoahmp.sh:1]: This file still contains some references to C768.

File: reg_tests/global_cycle/C192.lndincsoilnoahmp.sh [hunk 1]
```diff
@@ -22,13 +22,13 @@ export CDATE=2019073000
...
 
-export CASE=C768
+export CASE=C192
 
-export COMIN=$HOMEreg/input_data
-export FNTSFA=$COMIN/gdas.t00z.rtgssthr.grb
-export FNSNOA=$COMIN/gdas.t00z.snogrb_t1534.3072.1536
-export FNACNA=$COMIN/gdas.t00z.seaice.5min.blend.grb
-export NST_FILE=$COMIN/gdas.t00z.dtfanl.nc
+export COMIN=$HOMEreg/input_data_noahmp
```
Comment (GeorgeGayno-NOAA) [inline reg_tests/global_cycle/C192.lndincsoilnoahmp.sh:28]: You can remove these commented out lines.

File: reg_tests/global_cycle/C192.lndincsoilnoahmp.sh
Comment (ClaraDraper-NOAA) [reply reg_tests/global_cycle/C192.lndincsoilnoahmp.sh:1]: I've removed these, and cleaned up the rest of the comments / print statements.

File: reg_tests/global_cycle/C192.lndincsoilnoahmp.sh (hunk 1 above)
Comment (ClaraDraper-NOAA) [reply reg_tests/global_cycle/C192.lndincsoilnoahmp.sh:28]: Done.

**Comment by GeorgeGayno-NOAA** [comment]
I am getting Doxygen warnings in landincrements.f90 for undefined variables/arguments. I got rid of them as follows:

```
--- a/sorc/global_cycle.fd/land_increments.f90
+++ b/sorc/global_cycle.fd/land_increments.f90
@@ -18,7 +18,7 @@ module land_increments
                                                      !! copied from GFS_typedefs.F90

     ! control state for soil analysis:
-    integer, parameter       :: lsoil_incr=3 ! number of layers to add incrments to
+    integer, parameter       :: lsoil_incr=3 !< number of layers to add incrments to

     real, parameter          :: tfreez=273.16 !< con_t0c  in physcons
 contains
@@ -37,12 +37,17 @@ contains
  !! @param[inout] RLA Latitude on the cubed-sphere tile
  !! @param[inout] RLO Longitude on the cubed-sphere tile
  !! @param[inout] STC_STATE
+ !! @param[inout] SMC_STATE
+ !! @param[inout] SLC_STATE
+ !! @param[out] stc_updated
+ !! @param[out] slc_updated
  !! @param[in] SOILSNOW_TILE Land mask for increments on the cubed-sphere tile
  !! @param[in] SOILSNOW_FG_TILE First guess land mask for increments on the cubed-sphere tile
  !! @param[in] LENSFC Number of land points on a tile
  !! @param[in] LSOIL Number of soil layers
  !! @param[in] IDIM 'I' dimension of a tile
  !! @param[in] JDIM 'J' dimension of a tile
+ !! @param[in] lsm
  !! @param[in] MYRANK MPI rank number
  !!
  !! @author Clara Draper. @date March 2021
@@ -441,8 +446,8 @@ end subroutine calculate_landinc_mask
 !! appropriate max. values for SMC).
 !! Here: adjust (frozen) soil moisture to be consistent with changes in
 !! soil temperature from DA
-
-!> @param[in] lsm Integer code for the LSM
+!!
+!! @param[in] lsm Integer code for the LSM
 !! @param[in] isot Integer code for the soil type data set
 !! @param[in] ivegsrc Integer code for the vegetation type data set
 !! @param[in] lensfc Number of land points for this tile
@@ -453,6 +458,9 @@ end subroutine calculate_landinc_mask
 !! @param[in] stc_adj Analysis soil temperature states
 !! @param[inout] smc_adj Analysis soil moisture states
 !! @param[inout] slc_adj Analysis liquid soil moisture states
+!! @param[in] stc_updated
+!! @param[in] slc_updated
+!! @param[in] zsoil
 !! @author Clara Draper @date April 2021

```

**Comment by GeorgeGayno-NOAA** [comment]
The `global_cycle` unit test needs updates: 
https://github.com/NOAA-PSL/UFS_UTILS/tree/feature/AddNoahmpIncr/tests/global_cycle

**Comment by GeorgeGayno-NOAA** [comment]
I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.

**Comment by GeorgeGayno-NOAA** [comment]
> I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.

May I check these changes into your branch?

**Comment by ClaraDraper-NOAA** [comment]
> > I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.
> 
> May I check these changes into your branch?

Yes, please do! I'm about to look at the unit test now.

**Comment by GeorgeGayno-NOAA** [comment]
> > > I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.
> > 
> > 
> > May I check these changes into your branch?
> 
> Yes, please do! I'm about to look at the unit test now.

I don't have permission to push to your branch. The updated files (for Jet and Orion) are on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/global_cycle`

**Comment by ClaraDraper-NOAA** [comment]
> > > > I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.
> > > 
> > > 
> > > May I check these changes into your branch?
> > 
> > 
> > Yes, please do! I'm about to look at the unit test now.
> 
> I don't have permission to push to your branch. The updated files (for Jet and Orion) are on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/global_cycle`

I've pushed the updates to the jet and orion driver scripts.

**Comment by ClaraDraper-NOAA** [comment]
> I am getting Doxygen warnings in landincrements.f90 for undefined variables/arguments. I got rid of them as follows:
> 
> ```
> --- a/sorc/global_cycle.fd/land_increments.f90
> +++ b/sorc/global_cycle.fd/land_increments.f90
> @@ -18,7 +18,7 @@ module land_increments
>                                                       !! copied from GFS_typedefs.F90
> 
>      ! control state for soil analysis:
> -    integer, parameter       :: lsoil_incr=3 ! number of layers to add incrments to
> +    integer, parameter       :: lsoil_incr=3 !< number of layers to add incrments to
> 
>      real, parameter          :: tfreez=273.16 !< con_t0c  in physcons
>  contains
> @@ -37,12 +37,17 @@ contains
>   !! @param[inout] RLA Latitude on the cubed-sphere tile
>   !! @param[inout] RLO Longitude on the cubed-sphere tile
>   !! @param[inout] STC_STATE
> + !! @param[inout] SMC_STATE
> + !! @param[inout] SLC_STATE
> + !! @param[out] stc_updated
> + !! @param[out] slc_updated
>   !! @param[in] SOILSNOW_TILE Land mask for increments on the cubed-sphere tile
>   !! @param[in] SOILSNOW_FG_TILE First guess land mask for increments on the cubed-sphere tile
>   !! @param[in] LENSFC Number of land points on a tile
>   !! @param[in] LSOIL Number of soil layers
>   !! @param[in] IDIM 'I' dimension of a tile
>   !! @param[in] JDIM 'J' dimension of a tile
> + !! @param[in] lsm
>   !! @param[in] MYRANK MPI rank number
>   !!
>   !! @author Clara Draper. @date March 2021
> @@ -441,8 +446,8 @@ end subroutine calculate_landinc_mask
>  !! appropriate max. values for SMC).
>  !! Here: adjust (frozen) soil moisture to be consistent with changes in
>  !! soil temperature from DA
> -
> -!> @param[in] lsm Integer code for the LSM
> +!!
> +!! @param[in] lsm Integer code for the LSM
>  !! @param[in] isot Integer code for the soil type data set
>  !! @param[in] ivegsrc Integer code for the vegetation type data set
>  !! @param[in] lensfc Number of land points for this tile
> @@ -453,6 +458,9 @@ end subroutine calculate_landinc_mask
>  !! @param[in] stc_adj Analysis soil temperature states
>  !! @param[inout] smc_adj Analysis soil moisture states
>  !! @param[inout] slc_adj Analysis liquid soil moisture states
> +!! @param[in] stc_updated
> +!! @param[in] slc_updated
> +!! @param[in] zsoil
>  !! @author Clara Draper @date April 2021
> ```

I've pushed these changes.

**Comment by ClaraDraper-NOAA** [comment]
@GeorgeGayno-NOAA I've updated the unit test code so that the old test now correctly runs. I'll update it tomorrow to test my newly added features (for Noah-MP).

**Comment by GeorgeGayno-NOAA** [comment]
> I got the regression tests running on Jet and Orion. And I can run on WCOSS2 when the maintenance ends.

I ran the tests on WCOSS2. Find my script on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/global_cycle/driver.wcoss2.sh`

**Comment by GeorgeGayno-NOAA** [comment]
@ClaraDraper-NOAA Just a few final checks before I merge.

**Comment by ClaraDraper-NOAA** [comment]
Running to a meeting now, but I have one more update to push (with a new unit test). It's running now, so I'll push it in an hour.

**Comment by ClaraDraper-NOAA** [comment]
I just pushed the updated unit test - from my end, it's ready to merge.

**Review by GeorgeGayno-NOAA — APPROVED** [review]
Looks good.

**Pull request merged**